cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java

```diff
@@ -0,0 +1,228 @@
+
+  public static SourceControlMetadataStore create(StructuredTableContext context) {
+    return new SourceControlMetadataStore(context);
+  }
+
+  private SourceControlMetadataStore(StructuredTableContext context) {
+    this.context = context;
+  }
+
+  private StructuredTable getNamespaceSourceControlMetadataTable() {
+    try {
+      if (namespaceSourceControlMetadataTable == null) {
+        namespaceSourceControlMetadataTable = context.getTable(
+            StoreDefinition.NamespaceSourceControlMetadataStore.NAMESPACE_SOURCE_CONTROL_METADATA);
+      }
+    } catch (TableNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    return namespaceSourceControlMetadataTable;
+  }
+
+  /**
+   * Retrieves the source control metadata for the specified application ID in the namespace from
+   * {@code NamespaceSourceControlMetadata} table.
+   *
+   * @param appId {@link ApplicationId} for which the source control metadata is being retrieved.
+   * @return The {@link SourceControlMeta} associated with the application ID, or {@code null} if no
+   *         metadata is found.
+   * @throws IOException If it fails to read the metadata.
+   */
+  @Nullable
+  public SourceControlMeta getNamespaceSourceControlMeta(ApplicationId appId) throws IOException {
+    List<Field<?>> primaryKey = getPrimaryKey(appId);
+    StructuredTable table = getNamespaceSourceControlMetadataTable();
+    Optional<StructuredRow> row = table.read(primaryKey);
+    if (row.isPresent()) {
+      StructuredRow nonNullRow = row.get();
+      String specificationHash = nonNullRow
+          .getString(StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD);
+      String commitId = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD);
+      Long lastSynced = nonNullRow.getLong(
+          StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD);
+      if (specificationHash == null && commitId == null && lastSynced == 0) {
+        return null;
+      }
+      return new SourceControlMeta(specificationHash, commitId, Instant.ofEpochMilli(lastSynced));
+    }
+    return null;
+  }
+
+  /**
+   * Sets the source control metadata for the specified application ID in the namespace.  Source
+   * control metadata will be null when the application is deployed in the namespace. It will be
+   * non-null when the application is pulled from the remote repository and deployed in the
+   * namespace.
+   *
+   * @param appId             {@link ApplicationId} for which the source control metadata is being
+   *                          set.
+   * @param sourceControlMeta The {@link SourceControlMeta} to be set. Can be {@code null} if
+   *                          application is just deployed.
+   * @throws IOException If failed to write the data.
+   */
+  public void setNamespaceSourceControlMeta(ApplicationId appId,
+      @Nullable SourceControlMeta sourceControlMeta)
```

Comment:
ok this answers my question above. I don't think we want to allow null source control meta. When there is no source control information (like when an app is first deployed), the caller should avoid calling this method.

Reply:
In the Namespace Pipelines page, we need to show the sync status (SYNCED or UNSYNCED) and last modified of **all** the applications deployed in the namespace. Just to clarify, if an app has not been pulled/pushed we still need to show the status of the app as UNSYNCED in the page. 

If we don't add the source control information when the app is deployed, the data will split into two tables. We don't support **JOIN** operation. First we would have to read data from AppSpec table and then merge  it with data from namespace source control metadata table. The filtering (eg, filter on UNSYNCED sync status) , sorting, searching , pagination becomes difficult. Instead of doing filtering, searching, sorting in memory, we wanted it to happen at database level.

Reply:
I see, that makes sense. Can you add a comment in the code with this information?

Reply:
Sure, done